swift_browser_ui_frontend/src/common/download.js

```diff
@@ -0,0 +1,230 @@
+          });
+          this.chunkBuffer = [];
+        });
+        break;
+      } else {
+        this.offset = 0;
+        ({ value: chunk, done: readerDone } = await this.reader.read());
+        this.chunk = chunk;
+        this.readerDone = readerDone;
+      }
+    }
+    if (this.readerDone) {
+      if (this.chunk !== undefined) {
+        this.chunkBuffer = this.chunkBuffer.concat(
+          Array.from(
+            this.chunk.subarray(this.offset),
+          ),
+        );
+        this.chunk = undefined;
+      }
+
+      if (this.chunkBuffer.length > 0) {
+        navigator.serviceWorker.ready.then(reg => {
+          reg.active.postMessage({
+            cmd: "decryptChunk",
+            chunk: new Uint8Array(this.chunkBuffer),
+          });
+          this.chunkBuffer = [];
+        });
+      } else {
+        console.log("Telling the serviceWorker that decryption is done.");
+        navigator.serviceWorker.ready.then(reg => {
+          reg.active.postMessage({
+            cmd: "decryptionFinished",
+          });
+        });
+      }
+    }
+  }
+
+  getFileUrl() {
+    let fileURL = new URL(
+      `/download/${this.project}/${this.container}/${this.object}`,
+      document.location.origin,
+    );
+    fileURL.searchParams.append("project", this.active.id);
+    return fileURL;
+  }
+
+  async getHeader(pubkey) {
+    let upInfo = await getUploadEndpoint(
+      this.active.id,
+      this.project,
+      this.container,
+    );
+    this.$store.commit("setUploadInfo", upInfo);
+
+    let signatureUrl = new URL(`/sign/${60}`, document.location.origin);
+    signatureUrl.searchParams.append("path", this.whitelistPath);
+    let signed = await GET(signatureUrl);
+    signed = await signed.json();
+    let whitelistUrl = new URL(this.whitelistPath, this.endpoint);
+    whitelistUrl.searchParams.append("valid", signed.valid);
+    whitelistUrl.searchParams.append("signature", signed.signature);
+    whitelistUrl.searchParams.append("flavor", "crypt4gh");
+    whitelistUrl.searchParams.append(
+      "session",
+      this.$store.state.uploadInfo.id,
+    );
+    let resp = await PUT(whitelistUrl, pubkey);
+    let headerPath = `/header/${this.active.id}/${this.container}/${this.object}`;
+    signatureUrl = new URL(`/sign/${60}`, document.location.origin);
+    signatureUrl.searchParams.append("path", headerPath);
+    signed = await GET(signatureUrl);
+    signed = await signed.json();
+    let headerUrl = new URL(this.endpoint.concat(headerPath));
+    headerUrl.searchParams.append("valid", signed.valid);
+    headerUrl.searchParams.append("signature", signed.signature);
+    headerUrl.searchParams.append(
+      "session",
+      this.$store.state.uploadInfo.id,
+    );
+    resp = await GET(headerUrl);
+    let header = await resp.text();
+    header = Uint8Array.from(atob(header), c => c.charCodeAt(0));
+    addHeader(
+      header,
+      this.object.split("/").pop(),
```

Comment:
Is the object's header not being uploaded with full path as the name?

Reply:
It's not uploaded, but is added to the `ServiceWorker` file system. It's only using the file name for download session, as directory structure doesn't have any meaning for single files. For eventual batch downloads, either with a tarball or by mounting a working directory / files through `File API`, it'll use the full paths.